scripts/factory-common-install/libexec/app_current_device_vnc.sh

```diff
@@ -0,0 +1,167 @@
+  local systemd_service_name="${4?}"
+  shift 4
+
+  echo "Closing systemd process"
+  if ! ssh "${vnc_user}@${vnc_hostname}" -p "${device_ssh_port}" "systemctl -q --no-pager --user stop '${systemd_service_name}'" >/dev/null; then
+    err="$?"
+    1>&2 echo "Error: $FUNCNAME: $err"
+    return "$err"
+  fi
+}
+
+enter_vnc_as_user() {
+  local err
+  local vnc_passwd
+  local vnc_user="${1?}"
+  shift 1
+  
+  local systemd_service_name="nixos-secure-factory-x11"
+  local vnc_local_port="5900"
+  local vnc_remote_port="5900"
+  read_or_prompt_for_current_device__hostname "vnc_hostname"
+  read_or_prompt_for_current_device__ssh_port "device_ssh_port"
+
+  cleanup() {
+    local vnc_user="${1?}"
+    local vnc_hostname="${2?}"
+    local device_ssh_port="${3?}"
+    local vnc_remote_port="${4?}"
+    local systemd_service_name="${5?}"
+    shift 5
+
+    echo "Cleanup required"
+    if is_x11vnc_systemd_process_stopped "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"; then
+        echo "x11vnc systemd process is already closed"
+    else
+        echo "Closing x11vnc systemd process"
+        close_x11vnc_systemd_process "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"
+    fi
+
+    if is_x11vnc_pid_stopped "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${vnc_remote_port}"; then
+        echo "x11vnc pid is already stopped"
+    else
+        echo "Closing x11vnc pid"
+        exit_x11vnc "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}"
+    fi
+
+    return 0
+  }
+  trap "cleanup '${vnc_user}' '${vnc_hostname}' '${device_ssh_port}' '${vnc_remote_port}' '${systemd_service_name}'" EXIT SIGINT SIGQUIT
+
+  user_info="$(get_factory_info__user_full_name)<$(get_factory_info__user_email)>"
+  vnc_passwd="$(pwgen -cn 32 | head -n1)"
+
+  # Validation that the service that we are about to launch does not exist yet.
+  # if the service is alive we ask the user if it wants to kill the existing session.
+  echo " -> Making sure x11vnc is not aready started"
+  echo '"${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"'
+  if ! is_x11vnc_systemd_process_stopped "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${systemd_service_name}"; then
+    active_session_info=$(ssh "${vnc_user}@${vnc_hostname}" -p "${device_ssh_port}" "systemctl -q --no-pager --user status '${systemd_service_name}'") && true
+    echo "An active x11vnc session has been detected"
+    echo "Here is some information about the active session"
+    echo "${active_session_info}"
+    echo ""
+    read -n 1 -p 'Do you want to close the session? (Y)es or (N)o' user_input
+    echo ""
+    case $user_input in
+      [yY] )
+        echo "Okay, we will kill the existing session"
+        cleanup "${vnc_user}" "${vnc_hostname}" "${device_ssh_port}" "${vnc_remote_port}" "${systemd_service_name}"
+        # Sleep 5 seconds to avoid getting kicked out by the other script cleanup.
```

Comment:
Not sure I understand this. Why is that required?

Reply:
When the user decides to kill the existing session here's what happen
- The clean-up function is runned
  - If the systemd-service is up
    - We stop the service
  - If x11vnc pid is alive
    - We send the stop command

On the old connection, here's what happens:
- vncviewer closes and return an error code
- It execute the cleanup function.
  - What happens if the new user doesn't wait before launching the systemd-service/x11vnc.
    - The original service think the service are is own and tries to kill it.

It would probably be better to store the process pid of both

Reply:
Ok. As a clean fix seem quite involve, I think we should keep this workaround for the time being. It might be well advised to have a TODO comment which mention that a better approach is required with potential solution as ideas (also in comments).